### PR TITLE
Add a fake uname implementation to the Linux stdenv (that returns hardcoded values)

### PR DESCRIPTION
**Pull request by dezgeg**
This has two purity advantages:

- This makes `uname -m` return the processor architecture we're building
  for, not the architecture the build is running on. This mainly
  affects ARM, where we want to build armv6l packages on an armv7l
  package. The v7 processors are very capable of executing v6 code,
  but many build systems will look at `uname -m` and decide to do
  different things based on the result. Examples:
    - coreutils & openssl both add some ARMv7-specific FPU optimization
      flags to gcc's command line which causes the binaries to SIGILL
      on ARMv6 processors.
    - The emacs, perl & ruby builds  create machine-specific directories
      like `$out/libexec/emacs/25.1/armv6l-unknown-linux-gnueabihf`

  Both of these are now fixed. And no, unlike PER_LINUX32 (which
  makes `uname -m` report `i686` when run on x86_64) there is no kernel
  option to achieve the same thing for returning armv6l on a ARMv7.

- Some packages like Xorg capture the kernel version where the build
  was done (it prints "Build Operating System: Linux 4.4.45 x86_64"
  to the logs). This improves binary reproducibility of such packages.

I've added this just for Linux for now (as it ended up requiring tweaking the bootstrap stages to get it right) but in principle it could be added for Darwin as well if there's a use case.

cc @edolstra

### CODE REVIEW

**Comment by expipiplus1** [comment]
I was playing with a similar solution last week (https://github.com/expipiplus1/nixpkgs/blob/17c8652edbe164457e763d37cdaea03cd7251f91/pkgs/tools/misc/coreutils/uname-armv7l.patch) and discovered that (annoyingly) some packages use a system call to determine that machine instead of calling `uname`, the go 1.4 compiler is an example.

@grahamc suggested some systemtap magic to intercept the system call

@bennofs has suggested overriding the `uname` function in libc by using `LD_PRELOAD`

**Comment by edolstra** [comment]
I'm not really in favor of this. It seems a bit hacky and adds considerable complexity to stdenv. (E.g. you have to be careful to ensure that fake-uname appears before coreutils' uname.) 

An alternative would be to patch coreutils' uname to return a deterministic value based on some environment variable (say `NIX_UNAME=<sysname>:<machine>:...`). Also hacky but it would probably be just a few lines of code and wouldn't require duplicating all of uname's argument parsing. It wouldn't fix the `uname()` syscall, but please let's not do global `LD_PRELOAD` hacks or (shudder) systemtap magic just for `go`.

(Actually the environment variable hack could also be done in Glibc's `uname()`. Then it would work for everything except direct syscalls.)

**Comment by dezgeg** [comment]
The glibc idea sounds good, I will give that a shot...

**Comment by dezgeg** [comment]
Apparently that's not easy, as glibc build just auto-generates some wrappers from a system call list instead of having `uname()` written in C code...

**Comment by edolstra** [comment]
Oh, too bad :-(

**Comment by bennofs** [comment]
@dezgeg perhaps you can remove `uname` from the "list of auto-generated stuff" and add it manually?

**Comment by dezgeg** [comment]
That seems too complex and risky for very little gain. For Go you can set `GOHOSTARCH`, `GOARCH` and `GOARM` manually anyway to avoid the autodetection. And in fact you really have to set them to avoid impurities because the build system also tries to execute various floating point instructions and checks if a `SIGILL` was raised or no...

Coreutils seems a better place to hack.

**Comment by Ericson2314** [comment]
I'd blacklist the system call itself at build time for sandboxed builds. Autodetection during configuration is rarely a good idea, period. Go wouldn't use glibc to do the system call anyways, would it?

**Comment by dezgeg** [comment]
At least in the bootstrap Go `uname` is called only in shell scripts and some C code.

File: pkgs/build-support/fake-uname/default.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{ stdenv, lib, bash }:
+
+let unameTuple = [
+    /* kernel-name */       "Linux"
+    /* node-name */         "localhost"
+    /* kernel-release */    "4.4.0-not-a-real-version"
+    /* kernel-version */    "#1 SMP PREEMPT Thu Jan  1 00:00:01 GMT 1970"
+    /* machine */           (builtins.head (lib.splitString "-" stdenv.system))
```
Comment (Ericson2314) [inline pkgs/build-support/fake-uname/default.nix]: For the record, `hostPlatform.system` is probably a better way to get what you want. Makes which of the 3 platforms you want explicit, and during cross compilation the `stdenv.*` stuff incorrectly reflects the build platform.

**Comment by edolstra** [comment]
Not sure if blacklisting is a good idea, but we could use @aszlig's seccomp patch for Nix to make the uname syscall return a different value. This would compensate for the lack of a `personality()` for ARMv6.

**Comment by dezgeg** [comment]
If I remember correctly, besides optionally killing the process seccomp BPF filters can only affect the return value of a system call, not any values in userspace memory.

**Comment by aszlig** [comment]
@dezgeg: Well, you can via `SECCOMP_RET_TRAP` and `SECCOMP_RET_TRACE`, but that will be complicated as well, is very architecture specific and isn't supported by libseccomp.

For example this was my WIP handler for rewriting the `stat`-family of syscalls:

```c++
void rewriteStatHandler(int signo, siginfo_t *si, void *ctx) {
    ucontext_t *uc = reinterpret_cast<ucontext_t*>(ctx);

    if (signo != SIGSYS || si->si_code != SYS_SECCOMP || !uc)
        throw SysError("unexpected SIGSYS triggered");

    if (si->si_arch == SCMP_ARCH_X86_64) {
        long ret = syscall(si->si_syscall,
                           uc->uc_mcontext.gregs[REG_RDI],
                           uc->uc_mcontext.gregs[REG_RSI]);
        uc->uc_mcontext.gregs[REG_RAX] = ret;
        if (ret != 0) return;
        struct stat st;
        memcpy(&st, &uc->uc_mcontext.gregs[REG_RSI], sizeof(st));
        st.st_uid = 0;
        st.st_gid = 0;
        uc->uc_mcontext.gregs[REG_RSI] = reinterpret_cast<greg_t>(&st);
    }
}
```

So you can change any register with this, but because you're working on register-level you're also highly architecture dependant.

Another reason why it's probably not a good idea is that seccomp also comes with a bit of [overhead](https://wiki.tizen.org/wiki/Security:Seccomp#Performance_Analysis).

**Comment by Ericson2314** [comment]
By "blacklisting" I mean just fail the build. If killing the process is also easier / more performant, all the better!

**Comment by veprbl** [comment]
On darwin it would be really nice to have a fake ```sw_vers```. In my experience, missing it was leading to much build breakage.

**Comment by matthewbauer** [comment]
We actually have a real sw_vers in Darwin.DarwinTools but a fake one might be useful too.

**Comment by mmahut** [comment]
Are there any updates on this pull request, please?

**Comment by expipiplus1** [comment]
@mmahut, @dezgeg suggested (
https://github.com/NixOS/nix/pull/1916#issuecomment-387543970) that it's
convenient to use a VM to do these builds (in lieu of this PR). @dezgeg, do
you have a one liner you can share for this?

**Comment by misuzu** [comment]
There is this patch from Ubuntu folks and it works great: https://lists.ubuntu.com/archives/kernel-team/2016-January/068203.html

```nix
{ config, pkgs, ... }:
{
  boot.kernelPatches = [
    {
      name = "compat_uts_machine";
      patch = ./compat_uts_machine.patch;
    }
  ];
  boot.kernelParams = [
    "compat_uts_machine=armv7l"
  ];
}
```

```bash
[misuzu@oracle:~]$ uname -m
aarch64
[misuzu@oracle:~]$ linux32
[misuzu@oracle:~]$ uname -m
armv7l
```

**Comment by lucasew** [comment]
Is it still relevant?

**Comment by Mic92** [comment]
This PR seems stalled. Closing.

**Comment by Artturin** [comment]
A deterministic uname was finally added in https://github.com/NixOS/nixpkgs/pull/210102

**Pull request closed without merging**